src/Compiler/Microsoft.CodeAnalysis.Razor/src/SymbolExtensions.cs

```diff
@@ -13,33 +14,76 @@ internal static class SymbolExtensions
         }
 
-        // the simple name doesn't include generic type params, so skip over them in the full name if they are present
-        var fullNameEndIndex = fullName[^1] == '>' ? fullName.LastIndexOf('<') : fullName.Length;
-        var fullNameStartIndex = fullName.LastIndexOf('.', fullNameEndIndex - 1) + 1;
+        var fullNameSpan = fullName.AsSpan();
 
-        if (!fullName.AsSpan(fullNameStartIndex, fullNameEndIndex - fullNameStartIndex).Equals(symbol.Name.AsSpan(), StringComparison.Ordinal))
+        // The simple name doesn't include generic type params, so skip over them in the full name if they are present.
+        // UNDONE: This will need adjustment if this method should need to handle types overloaded on the number
+        // of generic type parameters.
+        if (fullNameSpan[^1] == '>')
         {
-            return false;
+            var endIndex = fullNameSpan.LastIndexOf('<');
+            Debug.Assert(endIndex >= 0, "Symbol name ends in '<' but does not contain '>'!");
+            fullNameSpan = fullNameSpan[..endIndex];
         }
 
-        // The simple name matched, do a fully qualified comparison
-        return symbol.ToDisplayString(FullNameTypeDisplayFormat).Equals(fullName, StringComparison.Ordinal);
+        // The strategy here is to walk backward through the various simple names in the full name and test
+        // them against the symbol parent chain.
+
+        for (ISymbol symbol = typeSymbol; !IsGlobalNamespace(symbol); symbol = symbol.ContainingSymbol)
+        {
+            var symbolNameSpan = symbol.Name.AsSpan();
+
+            // If the full name span doesn't end with the current symbol name, this isn't a match.
+            if (!fullNameSpan.EndsWith(symbolNameSpan, StringComparison.Ordinal))
+            {
+                return false;
+            }
+
+            fullNameSpan = fullNameSpan[..^symbolNameSpan.Length];
+
+            if (fullNameSpan.Length == 0)
+            {
+                // We just completed testing against full name.
+                // Check to see if the parent symbol is the global namespace. If it is,
+                // we're at the last symbol that we need to check and have a match!
+                // Otherwise, this match failed.
+                return IsGlobalNamespace(symbol.ContainingSymbol);
+            }
+
+            // Check the previous character. If it isn't a '.', we don't have a match.
+            // UNDONE: This will need adjustment if this method is expected to work properly with
+            // nested types using a '+' character.
+            if (fullNameSpan[^1] != '.')
+            {
+                return false;
+            }
+
+            // Skip the dot and loop to check the parent symbol.
+            fullNameSpan = fullNameSpan[..^1];
+        }
+
+        return false;
+
+        static bool IsGlobalNamespace(ISymbol symbol)
+        {
+            return symbol is INamespaceSymbol { IsGlobalNamespace: true };
+        }
```

Comment:
This code will be of particular interest, since @chsienki had made some performance improvements here already, and it's called in so many places. `ToDisplayString()` is no longer called here.